core/drand_test.go

```diff
@@ -37,228 +32,324 @@ func setFDLimit() {
+	assert.Equal(t, uint64(2), response.Round)
 }
 
-func TestDrandDKGBroadcastDeny(t *testing.T) {
+// Test dkg when two nodes cannot broadcast messages between them. The rest of the nodes
+// will be able to broadcast messages, so the process should finish successfully
+// Given 4 nodes = [ 0, 1, 2, 3]
+// 1. Limit communication between 1 and 2
+// 2. Run DKG
+// 3. Run reshare
+func TestRunDKGBroadcastDeny(t *testing.T) {
 	n := 4
 	thr := 3
 	beaconPeriod := 1 * time.Second
 
-	dt := NewDrandTest2(t, n, thr, beaconPeriod)
+	dt := NewDrandTestScenario(t, n, thr, beaconPeriod)
 	defer dt.Cleanup()
+
 	// close connection between a pair of nodes
-	n1 := dt.nodes[1]
-	n2 := dt.nodes[2]
-	n1.drand.DenyBroadcastTo(n2.addr)
-	n2.drand.DenyBroadcastTo(n1.addr)
+	node_1 := dt.nodes[1]
+	node_2 := dt.nodes[2]
+
+	t.Log("Setting node 1 not to broadcast messages to the node 2")
+	node_1.drand.DenyBroadcastTo(t, node_2.addr)
+
+	t.Log("Setting node 2 not to broadcast messages to the node 1")
+	node_2.drand.DenyBroadcastTo(t, node_1.addr)
+
 	group1 := dt.RunDKG()
-	dt.MoveToTime(group1.GenesisTime)
-	dt.MoveTime(1 * time.Second)
-	fmt.Printf("\n\n --- DKG FINISHED ---\n\n")
-	time.Sleep(200 * time.Millisecond)
-	_, err := dt.RunReshare(n, 0, thr, 1*time.Second, false, false)
+
+	// Advance clock
+	dt.SetMockClock(t, group1.GenesisTime)
+	dt.AdvanceMockClock(t, 1*time.Second)
+
+	group2, err := dt.RunReshare(t, nil, n, 0, thr, 1*time.Second, false, false, false)
 	require.NoError(t, err)
-	fmt.Println(" --- RESHARING FINISHED ---")
+	require.NotNil(t, group2)
+
+	t.Log("Resharing complete")
 }
 
-func TestDrandReshareForce(t *testing.T) {
-	oldN := 4
-	oldThr := 3
+// Test the dkg reshare can be forced to restart and finish successfully
+// when another dkg reshare was running before
+func TestRunDKGReshareForce(t *testing.T) {
+	oldNodes := 4
+	oldThreshold := 3
 	timeout := 1 * time.Second
 	beaconPeriod := 2 * time.Second
 
-	dt := NewDrandTest2(t, oldN, oldThr, beaconPeriod)
+	dt := NewDrandTestScenario(t, oldNodes, oldThreshold, beaconPeriod)
 	defer dt.Cleanup()
+
 	group1 := dt.RunDKG()
-	// make sure all nodes had enough time to run their go routines to start the
-	// beacon handler - related to CI problems
-	time.Sleep(getSleepDuration())
-	dt.MoveToTime(group1.GenesisTime)
-	dt.MoveTime(1 * time.Second)
+
+	// TODO: make sure all nodes had enough time to run their go routines to start the beacon handler - related to CI problems
+	// time.Sleep(getSleepDuration())
+
+	dt.SetMockClock(t, group1.GenesisTime)
+	dt.AdvanceMockClock(t, 1*time.Second)
 
 	// run the resharing
-	go dt.RunReshare(oldN, 0, oldThr, timeout, false, true)
-	time.Sleep(500 * time.Millisecond)
+	stateCh := make(chan int)
+	go func() {
+		t.Log("[ReshareForce] Start reshare")
+		_, err := dt.RunReshare(t, &stateCh, oldNodes, 0, oldThreshold, timeout, false, true, false)
+		require.Error(t, err)
+	}()
+
+	for {
+		if state := <-stateCh; state == ReshareUnlock {
```

Comment:
Nice ! Maybe add a `case <- time.After(2 * time.Minute)` or stg to break the test if nothing happens for a long time.

Reply:
Done!